# Answering Business Questions using SQL

We will use the chinook database for this project. The chinook database is provided as a SQLite database file called chinook.db.


![](chinook-schema.svg)

In [34]:
from sqlalchemy.engine import create_engine
import sqlite3
from pandas.io import sql
import subprocess
%reload_ext sql

In [35]:
%sql sqlite:///chinook.db

'Connected: @chinook.db'

## Overview of the Data

Let us start by getting familiar with the data

In [36]:
%%sql
SELECT 
     name,
     type
FROM sqlite_master
WHERE type IN ('table','view');

 * sqlite:///chinook.db
Done.


[('albums', 'table'),
 ('sqlite_sequence', 'table'),
 ('artists', 'table'),
 ('customers', 'table'),
 ('employees', 'table'),
 ('genres', 'table'),
 ('invoices', 'table'),
 ('invoice_items', 'table'),
 ('media_types', 'table'),
 ('playlists', 'table'),
 ('playlist_track', 'table'),
 ('tracks', 'table'),
 ('sqlite_stat1', 'table')]

## Selecting Albums to Purchase

The Chinook record store has just signed a deal with a new record label, and the task is selecting the first three albums that will be added to the store, from a list of four. All four albums are by artists that do not have any tracks in the store right now - we have the artist names, and the genre of the music they produce.
|Artist | Genre |
|:----------:|:----------:|
|Regal | Hip-hop|
|Red Tone | Punk|
|Meteor and The Girls | Pop|
|Slim Jim Bites |Blues| 

The record label specializes in artists from the USA, and they have given Chinook some money to advertise new albums in the USA, so we are interested in finding out which genres sell the best in the USA.

In [37]:
%%sql
WITH genre_sold_usa AS
(
    SELECT 
       g.Name genre,
       SUM(il.Quantity) num_tracks
      FROM genres g
      INNER JOIN tracks t on t.GenreId = g.GenreId
      INNER JOIN invoice_items il on il.TrackId = t.TrackId
      INNER JOIN invoices i on i.InvoiceId = il.InvoiceId
      INNER JOIN customers c on c.CustomerId = i.CustomerId
     WHERE c.country = 'USA'
     GROUP BY 1
     ORDER BY 2
)

SELECT 
    gsu.genre Genre,
    gsu.num_tracks abs_tracks
FROM genre_sold_usa gsu;

 * sqlite:///chinook.db
Done.


[('Science Fiction', 1),
 ('Easy Listening', 3),
 ('Rock And Roll', 3),
 ('Heavy Metal', 4),
 ('Hip Hop/Rap', 4),
 ('Soundtrack', 4),
 ('Alternative', 5),
 ('Pop', 5),
 ('Sci Fi & Fantasy', 5),
 ('Drama', 6),
 ('Reggae', 6),
 ('Bossa Nova', 7),
 ('Classical', 8),
 ('Comedy', 8),
 ('R&B/Soul', 12),
 ('TV Shows', 14),
 ('Blues', 15),
 ('Jazz', 22),
 ('Alternative & Punk', 50),
 ('Metal', 64),
 ('Latin', 91),
 ('Rock', 157)]

From the above, it can be said that Rock sells best in the USA, with a total of 157 tracks sold so far. However, Rock is not available in the record label.
Below is a list of the number of tracks sold according to the genres available in the record label, in a decreasing order, on the basis of no. of tracks sold.
1. Alternative & Punk - 50
2. Blues - 15
3. Pop - 5
4. Hip Hop/Rap - 4

Based on the list, we should purchase the albums of artists Red Tone, Slim Jim Bites, and Meteor and The Girls respectively, since their genres sell well.

## Analyzing Employee Sales Performace

Each customer for the Chinook store gets assigned to a sales support agent within the company when they first make a purchase. We have been asked to analyze the purchases of customers belonging to each employee to see if any sales support agent is performing either better or worse than the others.

In [38]:
%%sql

SELECT 
   e.FirstName || ' ' || e.LastName name,
   e.Title Title,
   e.BirthDate Birth_Date,
   e.HireDate Hire_Date,
   max(i.InvoiceDate) Latest_invoice_date,
   ROUND(SUM(i.Total), 2) amount_sales
  FROM employees e
  LEFT JOIN customers c on e.EmployeeId = c.SupportRepId
  LEFT JOIN invoices i on i.CustomerId = c.CustomerId

  WHERE Title = 'Sales Support Agent'
  GROUP BY 1
  ORDER BY 2

 * sqlite:///chinook.db
Done.


[('Jane Peacock', 'Sales Support Agent', '1973-08-29 00:00:00', '2002-04-01 00:00:00', '2013-12-22 00:00:00', 833.04),
 ('Margaret Park', 'Sales Support Agent', '1947-09-19 00:00:00', '2003-05-03 00:00:00', '2013-12-09 00:00:00', 775.4),
 ('Steve Johnson', 'Sales Support Agent', '1965-03-03 00:00:00', '2003-10-17 00:00:00', '2013-12-05 00:00:00', 720.16)]

## Analyzing Sales by Country

The next task is to analyze the sales data for customers from each country. The country value from customers table will be used for this analysis, and ignore the country from the billing address in the invoice table.

We should calculate data for each country on the
- total number of customers
- total value of sales
- average value of sales per customer
- average order value

Since there are a number of countries with only one customer, these customers should be grouped as "Other" in the analysis.

In [39]:
%%sql

WITH no_of_cstmrs AS
(
    SELECT 
       country Country,
       count(FirstName) No_of_customers
     FROM customers
      GROUP BY 1
      ORDER BY 2 DESC
)

SELECT * FROM no_of_cstmrs

 * sqlite:///chinook.db
Done.


[('USA', 13),
 ('Canada', 8),
 ('France', 5),
 ('Brazil', 5),
 ('Germany', 4),
 ('United Kingdom', 3),
 ('Portugal', 2),
 ('India', 2),
 ('Czech Republic', 2),
 ('Sweden', 1),
 ('Spain', 1),
 ('Poland', 1),
 ('Norway', 1),
 ('Netherlands', 1),
 ('Italy', 1),
 ('Ireland', 1),
 ('Hungary', 1),
 ('Finland', 1),
 ('Denmark', 1),
 ('Chile', 1),
 ('Belgium', 1),
 ('Austria', 1),
 ('Australia', 1),
 ('Argentina', 1)]

In [40]:
%%sql

WITH no_of_cstmrs AS
(
    SELECT 
       country Country,
       count(FirstName) No_of_customers
     FROM customers
      GROUP BY 1
      ORDER BY 2 DESC
),

total_sales_value AS
(SELECT c.country Country,
       ROUND(SUM(i.total), 2) Total_Value_Sales,
       ROUND(AVG(i.total), 2) Avg_Order_Value,
       ROUND(SUM(i.total)/(SELECT COUNT(DISTINCT(CustomerId)) FROM invoices), 2) Avg_Sales_Value_per_Customer
 
 FROM customers c
INNER JOIN invoices i ON i.CustomerId = c.CustomerId
GROUP BY 1
ORDER BY 2 DESC),

sales_data_1 AS
(SELECT noc.Country,
       noc.No_of_customers,
       tsv.Total_Value_Sales,
       tsv.Avg_Sales_Value_per_Customer,
       tsv.Avg_Order_Value
  FROM no_of_cstmrs noc  
INNER JOIN total_sales_value tsv ON tsv.Country=noc.Country
 WHERE noc.No_of_customers > 1),

sales_data_2 AS
(SELECT CASE
       WHEN noc.No_of_customers=1 THEN "Others"
       END AS
       Country,
       noc.No_of_customers,
       ROUND(AVG(tsv.Total_Value_Sales),2) Total_Value_Sales,
       ROUND(AVG(tsv.Avg_Sales_Value_per_Customer),2) AVG_Sales_Value_per_Customer,
       ROUND(AVG(tsv.Avg_Order_Value),2) AVG_Order_Value
  FROM no_of_cstmrs noc  
INNER JOIN total_sales_value tsv ON tsv.Country=noc.Country
 WHERE noc.No_of_customers = 1),

total_sales_data AS
(SELECT * 
FROM sales_data_1
UNION
SELECT * 
FROM sales_data_2)

SELECT *
  FROM total_sales_data
ORDER BY 3 DESC;

 * sqlite:///chinook.db
Done.


[('USA', 13, 523.06, 8.87, 5.75),
 ('Canada', 8, 303.96, 5.15, 5.43),
 ('France', 5, 195.1, 3.31, 5.57),
 ('Brazil', 5, 190.1, 3.22, 5.43),
 ('Germany', 4, 156.48, 2.65, 5.59),
 ('United Kingdom', 3, 112.86, 1.91, 5.37),
 ('Czech Republic', 2, 90.24, 1.53, 6.45),
 ('Portugal', 2, 77.24, 1.31, 5.52),
 ('India', 2, 75.26, 1.28, 5.79),
 ('Others', 1, 40.29, 0.68, 5.75)]

## Albums vs Individual Tracks

The Chinook store is setup in a way that allows customer to make purchases in one of the two ways:
- purchase a whole album
- purchase a collection of one or more individual tracks

The store does not let customers purchase a whole album, and then add individuals tracks to the same purchase (unless they do that by choosing each track manually). When customers purchase albums they are charged the same price as if they had purchased each of those tracks separately.

Management are currently considering changing their purchase strategy to save money. The strategy considered is to purchase only the most popular tracks from each album from record companies, instead of purchasing every track from an album.

We have to find out what percentage of purchases are individual tracks vs whole albums, so that management can use this data to understand the effect this decision might have on overall revenue.

It is quite common to have "edge cases" which prevent us from getting a 100% accurate answer to the question. In this instance, we have two edge cases to consider:
- Albums that have only one or two tracks are likely to be purchased by customers as part of a collection of individual tracks.
- Customers may decide to manually select every track from an album, and then add a few individual tracks from other albums to the purchase.

In the first case, since the analysis is concerned with maximizing revenue, we can safely ignore albums consisting of a few tracks. The company has previously done analysis to confirm that the second case does not happen often, so we can ignore this also.

In order to answer the question, we are going to have to identify whether each invoice has all the tracks from an album. We can do this by getting the list of tracks from an invoice and comparing it to the list of tracks from an album. We can find the album to compare the purchase to by looking up the album that one of the purchased tracks belongs to. It doesn't matter which track we pick, since if it's an album purchase, that album will be same for all tracks.

In [41]:
%%sql

SELECT max(InvoiceId) total_invoice
    FROM invoice_items;

 * sqlite:///chinook.db
Done.


[(412,)]

From the above table we observe that there are a total of 412 invoices. Further, we can compare the tracks from invoice with that of albums

In [42]:
%%sql

WITH com_prchs AS
(
    SELECT 
       il.InvoiceId invoice_id, 
       il.TrackId Prchs_tracks, 
       t.AlbumId album_id
      FROM tracks t
    INNER JOIN invoice_items il on il.TrackId = t.TrackId
    ORDER BY 1
)

SELECT *
 FROM com_prchs;

 * sqlite:///chinook.db
Done.


[(1, 2, 2),
 (1, 4, 3),
 (2, 6, 1),
 (2, 8, 1),
 (2, 10, 1),
 (2, 12, 1),
 (3, 16, 4),
 (3, 20, 4),
 (3, 24, 5),
 (3, 28, 5),
 (3, 32, 5),
 (3, 36, 5),
 (4, 42, 6),
 (4, 48, 6),
 (4, 54, 7),
 (4, 60, 7),
 (4, 66, 8),
 (4, 72, 8),
 (4, 78, 9),
 (4, 84, 9),
 (4, 90, 10),
 (5, 99, 11),
 (5, 108, 11),
 (5, 117, 12),
 (5, 126, 13),
 (5, 135, 14),
 (5, 144, 15),
 (5, 153, 16),
 (5, 162, 17),
 (5, 171, 18),
 (5, 180, 18),
 (5, 189, 19),
 (5, 198, 20),
 (5, 207, 21),
 (5, 216, 21),
 (6, 230, 23),
 (7, 231, 23),
 (7, 232, 23),
 (8, 234, 23),
 (8, 236, 23),
 (9, 238, 23),
 (9, 240, 23),
 (9, 242, 23),
 (9, 244, 23),
 (10, 248, 24),
 (10, 252, 24),
 (10, 256, 24),
 (10, 260, 24),
 (10, 264, 24),
 (10, 268, 24),
 (11, 274, 25),
 (11, 280, 25),
 (11, 286, 26),
 (11, 292, 26),
 (11, 298, 26),
 (11, 304, 27),
 (11, 310, 27),
 (11, 316, 28),
 (11, 322, 28),
 (12, 331, 29),
 (12, 340, 30),
 (12, 349, 30),
 (12, 358, 31),
 (12, 367, 32),
 (12, 376, 33),
 (12, 385, 33),
 (12, 394, 34),
 (12, 403, 34),
 (12, 412, 35),
 (12, 421, 36),
 (12, 430, 36),
 (12, 439, 37),
 (12, 448, 37),
 (13, 462, 38),
 (14, 463, 38),
 (14, 464, 38),
 (15, 466, 38),
 (15, 468, 39),
 (16, 470, 39),
 (16, 472, 39),
 (16, 474, 39),
 (16, 476, 39),
 (17, 480, 39),
 (17, 484, 39),
 (17, 488, 39),
 (17, 492, 40),
 (17, 496, 40),
 (17, 500, 40),
 (18, 506, 41),
 (18, 512, 41),
 (18, 518, 23),
 (18, 524, 23),
 (18, 530, 42),
 (18, 536, 42),
 (18, 542, 42),
 (18, 548, 43),
 (18, 554, 44),
 (19, 563, 45),
 (19, 572, 46),
 (19, 581, 46),
 (19, 590, 47),
 (19, 599, 48),
 (19, 608, 48),
 (19, 617, 49),
 (19, 626, 51),
 (19, 635, 51),
 (19, 644, 51),
 (19, 653, 52),
 (19, 662, 53),
 (19, 671, 53),
 (19, 680, 54),
 (20, 694, 54),
 (21, 695, 55),
 (21, 696, 55),
 (22, 698, 55),
 (22, 700, 55),
 (23, 702, 55),
 (23, 704, 55),
 (23, 706, 55),
 (23, 708, 55),
 (24, 712, 55),
 (24, 716, 56),
 (24, 720, 56),
 (24, 724, 56),
 (24, 728, 56),
 (24, 732, 57),
 (25, 738, 57),
 (25, 744, 57),
 (25, 750, 58),
 (25, 756, 59),
 (25, 762, 60),
 (25, 768, 61),
 (25, 774, 61),
 (25, 780, 62),
 (25, 786, 63),
 (26, 795, 63),
 (26, 804, 64),
 (26, 813, 65),
 (26, 822, 66),
 (26, 831, 67),
 (26, 840, 67),
 (26, 849, 68),
 (26, 858, 69),
 (26, 867, 70),
 (26, 876, 70),
 (26, 885, 71),
 (26, 894, 72),
 (26, 903, 72),
 (26, 912, 73),
 (27, 926, 74),
 (28, 927, 74),
 (28, 928, 74),
 (29, 930, 74),
 (29, 932, 74),
 (30, 934, 74),
 (30, 936, 75),
 (30, 938, 75),
 (30, 940, 75),
 (31, 944, 75),
 (31, 948, 75),
 (31, 952, 76),
 (31, 956, 76),
 (31, 960, 76),
 (31, 964, 77),
 (32, 970, 77),
 (32, 976, 78),
 (32, 982, 78),
 (32, 988, 78),
 (32, 994, 79),
 (32, 1000, 80),
 (32, 1006, 80),
 (32, 1012, 81),
 (32, 1018, 81),
 (33, 1027, 82),
 (33, 1036, 83),
 (33, 1045, 83),
 (33, 1054, 83),
 (33, 1063, 84),
 (33, 1072, 84),
 (33, 1081, 85),
 (33, 1090, 86),
 (33, 1099, 86),
 (33, 1108, 73),
 (33, 1117, 73),
 (33, 1126, 88),
 (33, 1135, 89),
 (33, 1144, 89),
 (34, 1158, 91),
 (35, 1159, 91),
 (35, 1160, 91),
 (36, 1162, 91),
 (36, 1164, 91),
 (37, 1166, 91),
 (37, 1168, 91),
 (37, 1170, 91),
 (37, 1172, 91),
 (38, 1176, 92),
 (38, 1180, 92),
 (38, 1184, 92),
 (38, 1188, 93),
 (38, 1192, 93),
 (38, 1196, 93),
 (39, 1202, 94),
 (39, 1208, 94),
 (39, 1214, 95),
 (39, 1220, 95),
 (39, 1226, 96),
 (39, 1232, 96),
 (39, 1238, 97),
 (39, 1244, 97),
 (39, 1250, 98),
 (40, 1259, 99),
 (40, 1268, 100),
 (40, 1277, 101),
 (40, 1286, 101),
 (40, 1295, 102),
 (40, 1304, 102),
 (40, 1313, 103),
 (40, 1322, 104),
 (40, 1331, 105),
 (40, 1340, 106),
 (40, 1349, 107),
 (40, 1358, 108),
 (40, 1367, 109),
 (40, 1376, 110),
 (41, 1390, 112),
 (42, 1391, 112),
 (42, 1392, 112),
 (43, 1394, 112),
 (43, 1396, 113),
 (44, 1398, 113),
 (44, 1400, 113),
 (44, 1402, 113),
 (44, 1404, 113),
 (45, 1408, 114),
 (45, 1412, 114),
 (45, 1416, 115),
 (45, 1420, 115),
 (45, 1424, 115),
 (45, 1428, 115),
 (46, 1434, 116),
 (46, 1440, 116),
 (46, 1446, 117),
 (46, 1452, 117),
 (46, 1458, 118),
 (46, 1464, 118),
 (46, 1470, 119),
 (46, 1476, 119),
 (46, 1

From the above table it can be observed that if the tracks belonging to each distinct invoice belongs to the same album it is an Album Purchase. If these tracks belong to different albums it is Non Albu Purchase.

Below, we can categorize these invoices into Album and Non Album Purchase. This can be done by counting the number of distinct albums belonging to each invoice. Then we compare the Distinct Albums and Number of Albums belonging to each invoice.

In [44]:
%%sql

WITH com_prchs AS
(
    SELECT 
       il.InvoiceId invoice_id, 
       il.TrackId Prchs_tracks, 
       t.AlbumId album_id
      FROM tracks t
    INNER JOIN invoice_items il on il.TrackId = t.TrackId
    ORDER BY 1
)

SELECT 
   invoice_id, 
   COUNT(DISTINCT(album_id)) Distinct_album_id,
   COUNT(album_id) No_of_album_id,
   CASE
   WHEN COUNT(album_id) - COUNT(DISTINCT(album_id)) <= 2 THEN "Non Album Purchase"
   ELSE "Album Purchase"
   END AS Category
  FROM com_prchs
 GROUP BY 1;

 * sqlite:///chinook.db
Done.


[(1, 2, 2, 'Non Album Purchase'),
 (2, 1, 4, 'Album Purchase'),
 (3, 2, 6, 'Album Purchase'),
 (4, 5, 9, 'Album Purchase'),
 (5, 11, 14, 'Album Purchase'),
 (6, 1, 1, 'Non Album Purchase'),
 (7, 1, 2, 'Non Album Purchase'),
 (8, 1, 2, 'Non Album Purchase'),
 (9, 1, 4, 'Album Purchase'),
 (10, 1, 6, 'Album Purchase'),
 (11, 4, 9, 'Album Purchase'),
 (12, 9, 14, 'Album Purchase'),
 (13, 1, 1, 'Non Album Purchase'),
 (14, 1, 2, 'Non Album Purchase'),
 (15, 2, 2, 'Non Album Purchase'),
 (16, 1, 4, 'Album Purchase'),
 (17, 2, 6, 'Album Purchase'),
 (18, 5, 9, 'Album Purchase'),
 (19, 9, 14, 'Album Purchase'),
 (20, 1, 1, 'Non Album Purchase'),
 (21, 1, 2, 'Non Album Purchase'),
 (22, 1, 2, 'Non Album Purchase'),
 (23, 1, 4, 'Album Purchase'),
 (24, 3, 6, 'Album Purchase'),
 (25, 7, 9, 'Non Album Purchase'),
 (26, 11, 14, 'Album Purchase'),
 (27, 1, 1, 'Non Album Purchase'),
 (28, 1, 2, 'Non Album Purchase'),
 (29, 1, 2, 'Non Album Purchase'),
 (30, 2, 4, 'Non Album Purchase'),
 (31, 3, 6, 'Album Purchase'),
 (32, 5, 9, 'Album Purchase'),
 (33, 8, 14, 'Album Purchase'),
 (34, 1, 1, 'Non Album Purchase'),
 (35, 1, 2, 'Non Album Purchase'),
 (36, 1, 2, 'Non Album Purchase'),
 (37, 1, 4, 'Album Purchase'),
 (38, 2, 6, 'Album Purchase'),
 (39, 5, 9, 'Album Purchase'),
 (40, 12, 14, 'Non Album Purchase'),
 (41, 1, 1, 'Non Album Purchase'),
 (42, 1, 2, 'Non Album Purchase'),
 (43, 2, 2, 'Non Album Purchase'),
 (44, 1, 4, 'Album Purchase'),
 (45, 2, 6, 'Album Purchase'),
 (46, 5, 9, 'Album Purchase'),
 (47, 11, 14, 'Album Purchase'),
 (48, 1, 1, 'Non Album Purchase'),
 (49, 1, 2, 'Non Album Purchase'),
 (50, 2, 2, 'Non Album Purchase'),
 (51, 2, 4, 'Non Album Purchase'),
 (52, 3, 6, 'Album Purchase'),
 (53, 4, 9, 'Album Purchase'),
 (54, 10, 14, 'Album Purchase'),
 (55, 1, 1, 'Non Album Purchase'),
 (56, 1, 2, 'Non Album Purchase'),
 (57, 1, 2, 'Non Album Purchase'),
 (58, 1, 4, 'Album Purchase'),
 (59, 3, 6, 'Album Purchase'),
 (60, 5, 9, 'Album Purchase'),
 (61, 9, 14, 'Album Purchase'),
 (62, 1, 1, 'Non Album Purchase'),
 (63, 1, 2, 'Non Album Purchase'),
 (64, 1, 2, 'Non Album Purchase'),
 (65, 4, 4, 'Non Album Purchase'),
 (66, 2, 6, 'Album Purchase'),
 (67, 5, 9, 'Album Purchase'),
 (68, 10, 14, 'Album Purchase'),
 (69, 1, 1, 'Non Album Purchase'),
 (70, 1, 2, 'Non Album Purchase'),
 (71, 1, 2, 'Non Album Purchase'),
 (72, 1, 4, 'Album Purchase'),
 (73, 2, 6, 'Album Purchase'),
 (74, 4, 9, 'Album Purchase'),
 (75, 10, 14, 'Album Purchase'),
 (76, 1, 1, 'Non Album Purchase'),
 (77, 1, 2, 'Non Album Purchase'),
 (78, 1, 2, 'Non Album Purchase'),
 (79, 1, 4, 'Album Purchase'),
 (80, 3, 6, 'Album Purchase'),
 (81, 4, 9, 'Album Purchase'),
 (82, 9, 14, 'Album Purchase'),
 (83, 1, 1, 'Non Album Purchase'),
 (84, 1, 2, 'Non Album Purchase'),
 (85, 1, 2, 'Non Album Purchase'),
 (86, 1, 4, 'Album Purchase'),
 (87, 3, 6, 'Album Purchase'),
 (88, 4, 9, 'Album Purchase'),
 (89, 8, 14, 'Album Purchase'),
 (90, 1, 1, 'Non Album Purchase'),
 (91, 1, 2, 'Non Album Purchase'),
 (92, 1, 2, 'Non Album Purchase'),
 (93, 2, 4, 'Non Album Purchase'),
 (94, 3, 6, 'Album Purchase'),
 (95, 5, 9, 'Album Purchase'),
 (96, 9, 14, 'Album Purchase'),
 (97, 1, 1, 'Non Album Purchase'),
 (98, 1, 2, 'Non Album Purchase'),
 (99, 2, 2, 'Non Album Purchase'),
 (100, 1, 4, 'Album Purchase'),
 (101, 2, 6, 'Album Purchase'),
 (102, 4, 9, 'Album Purchase'),
 (103, 11, 14, 'Album Purchase'),
 (104, 1, 1, 'Non Album Purchase'),
 (105, 2, 2, 'Non Album Purchase'),
 (106, 2, 2, 'Non Album Purchase'),
 (107, 4, 4, 'Non Album Purchase'),
 (108, 4, 6, 'Non Album Purchase'),
 (109, 5, 9, 'Album Purchase'),
 (110, 10, 14, 'Album Purchase'),
 (111, 1, 1, 'Non Album Purchase'),
 (112, 1, 2, 'Non Album Purchase'),
 (113, 1, 2, 'Non Album Purchase'),
 (114, 1, 4, 'Album Purchase'),
 (115, 2, 6, 'Album Purchase'),
 (116, 4, 9, 'Album Purchase'),
 (117, 10, 14, 'Album Purchase'),
 (118, 1, 1, 'Non Album Purchase'),
 (119, 1, 2, 'Non Album Purchase'),
 (120, 1, 2, 'Non Album Purchase'),
 

### Category: Not an Album Purchase

Here we calculate the Percentage of Non Album Purchases for the company.

In [51]:
%%sql

WITH com_prchs AS
(SELECT il.InvoiceId invoice_id, il.TrackId Pur_tracks, t.AlbumId album_id
  FROM tracks t
INNER JOIN invoice_items il ON il.TrackId = t.TrackId
ORDER BY 1
),

ctgr_prchs AS
(SELECT invoice_id, COUNT(DISTINCT(album_id)) Distinct_album_id, 
       COUNT(album_id) No_of_album_id,
       CASE
       WHEN COUNT(album_id) - COUNT(DISTINCT(album_id)) <= 2 THEN "Non Album Purchase"
       ELSE "Album Purchase"
       END AS Category 
  FROM com_prchs
GROUP BY 1
)

SELECT (SELECT MAX(InvoiceId)
          FROM invoice_items) TOTAL_invoice_id,
       COUNT(invoice_id) No_of_invoice_id,
       (SELECT ROUND(CAST(COUNT(invoice_id) AS FLOAT) * 100 / (SELECT MAX(invoice_id)
                                            FROM invoice_items), 2)) Perc_invoice
  FROM ctgr_prchs
WHERE Category = "Non Album Purchase";

 * sqlite:///chinook.db
Done.


[(412, 213, 51.7)]

From the above table, it can be seen that nearly half of the purchases, i.e., 213 out of 412 are Non Album Purchase.

### Category: Album Purchase

Here we calculate the Percentage of Album Purchase for the company.

In [53]:
%%sql

WITH com_prchs AS
(SELECT il.InvoiceId invoice_id, il.TrackId Pur_tracks, t.AlbumId album_id
  FROM tracks t
INNER JOIN invoice_items il ON il.TrackId = t.TrackId
ORDER BY 1
),

ctgr_prchs AS
(SELECT invoice_id, COUNT(DISTINCT(album_id)) Distinct_album_id, 
       COUNT(album_id) No_of_album_id,
       CASE
       WHEN COUNT(album_id) - COUNT(DISTINCT(album_id)) <= 2 THEN "Non Album Purchase"
       ELSE "Album Purchase"
       END AS Category 
  FROM com_prchs
GROUP BY 1
)

SELECT (SELECT MAX(InvoiceId)
          FROM invoice_items) TOTAL_invoice_id,
       COUNT(invoice_id) No_of_invoice_id,
       (SELECT ROUND(CAST(COUNT(invoice_id) AS FLOAT) * 100 / (SELECT MAX(invoice_id)
                                            FROM invoice_items), 2)) Perc_invoice
  FROM ctgr_prchs
WHERE Category = "Album Purchase";

 * sqlite:///chinook.db
Done.


[(412, 199, 48.42)]

Here, it can be seen that nearly 48%, i.e., 199 out of 412 purchase are Album Purchases.

From the above two intepretations, it can be seen that Non Album Purchases slightly sell better. The store should consider buying popular tracks from albums rather than buying the whole album.